tools/java/org/unicode/cldr/util/XPathParts.java

```diff
@@ -439,7 +440,18 @@ public String getAttributeValue(int elementIndex, String attribute) {
     }
 
     public void putAttributeValue(int elementIndex, String attribute, String value) {
-        elements.get(elementIndex >= 0 ? elementIndex : elementIndex + size()).putAttribute(attribute, value);
+        elementIndex = elementIndex >= 0 ? elementIndex : elementIndex + size();
+        Map<String, String> ea = elements.get(elementIndex).attributes;
+        if (value == null && (ea == null || !ea.containsKey(attribute))) {
+            return;
+        }
+        if (value != null && ea != null && ea.containsKey(attribute) && ea.get(attribute).equals(value)) {
```

Comment:
ea.containsKey(attribute) && ea.get(attribute).equals(value)

should be optimized (since value is at this point guaranteed not to be null) as:

value.equals(ea.get(attribute))